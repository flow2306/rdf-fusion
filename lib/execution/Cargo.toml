[package]
name = "rdf-fusion-execution"
version.workspace = true
authors.workspace = true
license.workspace = true
edition.workspace = true
rust-version.workspace = true
description = """
RDF Fusion Execution Engine
"""

[dependencies]
async-trait.workspace = true
datafusion.workspace = true
itertools.workspace = true
rdf-fusion-extensions.workspace = true
rdf-fusion-encoding.workspace = true
rdf-fusion-functions.workspace = true
rdf-fusion-model.workspace = true
futures.workspace = true
rdf-fusion-logical.workspace = true
rdf-fusion-physical.workspace = true
oxrdfio.workspace = true
spargebra.workspace = true
sparesults.workspace = true
thiserror.workspace = true
tokio.workspace = true

[lints]
workspace = true
