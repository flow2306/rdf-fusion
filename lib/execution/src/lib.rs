#![doc(test(attr(deny(warnings))))]
#![doc(
    html_favicon_url = "https://raw.githubusercontent.com/tobixdev/rdf-fusion/main/misc/logo/logo.png"
)]
#![doc(
    html_logo_url = "https://raw.githubusercontent.com/tobixdev/rdf-fusion/main/misc/logo/logo.png"
)]

//! This crate defines the execution engine of RDF Fusion.
//!
//! # RDF Fusion Context
//!
//! Similar to DataFusion’s [`SessionContext`](datafusion::prelude::SessionContext),
//! RDF Fusion provides an [`RdfFusionContext`].
//! This context manages the state of the RDF Fusion engine, including the registered RDF term
//! encodings.
//! For more details, see the [`RdfFusionContext`] documentation.
//!
//! # Executing SPARQL Queries
//!
//! DataFusion operates on the concept of a *logical plan*.
//! While DataFusion includes an SQL front-end, it does not provide a SPARQL front-end.
//! To execute SPARQL queries, we convert them into logical plans through the following pipeline:
//!
//! ```text
//! Query String -> SPARQL Algebra (Oxigraph) -> Logical Plan
//! ```
//!
//! The query string is first parsed by Oxigraph’s SPARQL parser, which produces a SPARQL algebra
//! graph pattern. This graph pattern already includes high-level concepts such as `Join`, `Filter`,
//! and `Projection`.
//!
//! We then convert this algebra pattern into a DataFusion logical plan through a rewriting step.
//! The resulting logical plan is executed by DataFusion.

extern crate core;

mod engine;
mod planner;
pub mod results;
pub mod sparql;

pub use engine::RdfFusionContext;
