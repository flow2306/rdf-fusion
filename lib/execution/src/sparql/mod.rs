//! [SPARQL](https://www.w3.org/TR/sparql11-overview/) implementation.

mod algebra;
pub mod error;
mod eval;
mod explanation;
mod optimizer;
mod rewriting;

pub use crate::sparql::algebra::{Query, QueryDataset, Update};
pub use crate::sparql::explanation::QueryExplanation;
pub use eval::evaluate_query;
pub use optimizer::{create_optimizer_rules, create_pyhsical_optimizer_rules};
pub use rdf_fusion_model::{Variable, VariableNameParseError};
pub use spargebra::SparqlSyntaxError;

/// Defines how many optimizations the query optimizer should apply.
///
/// Currently, the default value is [OptimizationLevel::Full], as we are still searching for a
/// subset that performs well on many queries. Once this subset has been identified, the default
/// value will be [OptimizationLevel::Default].
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub enum OptimizationLevel {
    /// No optimizations, except rewrites that are necessary for a working query.
    None,
    /// A balanced default optimization level. Suitable for simple queries or those handling modest
    /// data volumes.
    Default,
    /// Runs all optimizations. Ideal for complex queries or those processing large datasets.
    #[default]
    Full,
}

/// Options for SPARQL query evaluation.
#[derive(Clone, Default)]
pub struct QueryOptions {
    /// The defined optimization level
    pub optimization_level: OptimizationLevel,
}

/// Options for SPARQL update evaluation.
#[derive(Clone, Default)]
pub struct UpdateOptions;

impl From<QueryOptions> for UpdateOptions {
    #[inline]
    fn from(_query_options: QueryOptions) -> Self {
        Self {}
    }
}
