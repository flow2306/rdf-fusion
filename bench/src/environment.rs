use crate::BenchmarkingOptions;
use crate::benchmarks::BenchmarkName;
use crate::prepare::{PrepRequirement, prepare_run_closure, prepare_run_command};
use crate::prepare::{ensure_file_download, prepare_file_download};
use anyhow::bail;
use datafusion::execution::runtime_env::{RuntimeEnv, RuntimeEnvBuilder};
use datafusion::prelude::SessionConfig;
use rdf_fusion::store::Store;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};

/// Represents a context used to execute benchmarks.
pub struct RdfFusionBenchContext {
    /// General options for the benchmarks.
    options: BenchmarkingOptions,
    /// The path to the data dir.
    data_dir: Mutex<PathBuf>,
    /// The path to the results dir.
    results_dir: Mutex<PathBuf>,
}

impl RdfFusionBenchContext {
    /// Creates a new [RdfFusionBenchContext].
    pub fn new(
        options: BenchmarkingOptions,
        data_dir: PathBuf,
        results_dir: PathBuf,
    ) -> Self {
        Self {
            options,
            data_dir: Mutex::new(data_dir),
            results_dir: Mutex::new(results_dir),
        }
    }

    /// Creates a new [RdfFusionBenchContext] used in the criterion benchmarks.
    pub fn new_for_criterion(data_dir: PathBuf, target_partitions: usize) -> Self {
        Self {
            options: BenchmarkingOptions {
                verbose_results: false,
                target_partitions: Some(target_partitions),
                memory_size: None,
            },
            data_dir: Mutex::new(data_dir),
            results_dir: Mutex::new(PathBuf::from("/temp")),
        }
    }

    /// Returns the [BenchmarkingOptions] for this context.
    pub fn options(&self) -> &BenchmarkingOptions {
        &self.options
    }

    /// Resolves a relative path `file` against the data directory.
    pub fn join_data_dir(&self, file: &Path) -> anyhow::Result<PathBuf> {
        if !file.is_relative() {
            bail!("Only relative paths can be resolved.")
        }

        Ok(self.data_dir.lock().unwrap().join(file))
    }

    pub fn create_store(&self) -> Store {
        let mut config = SessionConfig::new()
            .with_batch_size(8192)
            .with_target_partitions(self.options.target_partitions.unwrap_or(1));

        let options = config.options_mut();
        options.optimizer.enable_dynamic_filter_pushdown = true;

        let runtime_enc = match self.options.memory_size {
            None => Arc::new(RuntimeEnv::default()),
            Some(memory_size) => RuntimeEnvBuilder::new()
                .with_memory_limit(memory_size * 1024 * 1024, 1f64)
                .build_arc()
                .expect("Only setting memory limit"),
        };
        Store::new_with_datafusion_config(config, runtime_enc)
    }

    /// Creates a new folder in the results directory and uses it until [Self::pop_dir] is
    /// called.
    ///
    /// This can be used to create folder hierarchies to separate the results of different
    /// benchmarks.
    #[allow(clippy::create_dir)]
    #[allow(clippy::unwrap_used, reason = "Mutex poisoning")]
    pub fn push_dir(
        &self,
        data_dir_name: &str,
        results_dir_name: &str,
    ) -> anyhow::Result<()> {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.push(data_dir_name);
        results_dir.push(results_dir_name);

        Ok(())
    }

    /// Pops the last directory from the stack.
    pub fn pop_dir(&self) {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.pop();
        results_dir.pop();
    }

    /// Creates a new bencher and modifies the context for this benchmark.
    pub fn create_benchmark_context(
        &self,
        benchmark_name: BenchmarkName,
    ) -> anyhow::Result<BenchmarkContext<'_>> {
        self.push_dir(
            &benchmark_name.data_dir_name(),
            &benchmark_name.results_dir_name(),
        )?;
        Ok(BenchmarkContext {
            context: self,
            benchmark_name,
        })
    }
}

/// A benchmarker that can be used to execute benchmarks.
///
/// It holds a reference to the current context to store its results.
pub struct BenchmarkContext<'ctx> {
    /// Reference to the benchmarking context.
    context: &'ctx RdfFusionBenchContext,
    /// Name of the benchmark that is being executed.
    benchmark_name: BenchmarkName,
}

impl<'ctx> BenchmarkContext<'ctx> {
    /// Returns a reference to the benchmarking context.
    pub fn parent(&self) -> &RdfFusionBenchContext {
        self.context
    }

    /// Returns the name of the benchmark that is being executed.
    pub fn benchmark_name(&self) -> BenchmarkName {
        self.benchmark_name
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn data_dir(&self) -> PathBuf {
        self.context.data_dir.lock().unwrap().clone()
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn results_dir(&self) -> PathBuf {
        self.context.results_dir.lock().unwrap().clone()
    }

    /// Prepares the context such that `requirement` is fulfilled.
    pub async fn prepare_requirement(
        &self,
        requirement: PrepRequirement,
    ) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload {
                url,
                file_name,
                action,
            } => prepare_file_download(self, url, file_name, action).await,
            PrepRequirement::RunClosure { execute, .. } => {
                prepare_run_closure(self, &execute)
            }
            PrepRequirement::RunCommand {
                workdir,
                program,
                args,
                ..
            } => {
                let workdir = self.context.join_data_dir(&workdir)?;
                prepare_run_command(&workdir, &program, &args)
            }
        }
    }

    /// Ensures that the `requirement` is fulfilled in this context.
    pub fn ensure_requirement(&self, requirement: PrepRequirement) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload { file_name, .. } => {
                ensure_file_download(self, file_name.as_path())
            }
            PrepRequirement::RunClosure {
                check_requirement, ..
            }
            | PrepRequirement::RunCommand {
                check_requirement, ..
            } => check_requirement(self),
        }
    }
}

/// Pops the results directory from the context when the bencher is dropped.
impl Drop for BenchmarkContext<'_> {
    fn drop(&mut self) {
        self.context.pop_dir();
    }
}
