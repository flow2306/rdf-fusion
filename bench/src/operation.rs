use crate::runs::BenchmarkRun;
use futures::StreamExt;
use rdf_fusion::execution::results::QueryResults;
use rdf_fusion::execution::sparql::{Query, QueryExplanation, QueryOptions};
use rdf_fusion::store::Store;

#[derive(Clone)]
pub enum SparqlRawOperation<QueryName> {
    Query(QueryName, String),
}

impl<QueryName: Clone> SparqlRawOperation<QueryName> {
    pub fn query_name(&self) -> QueryName {
        match self {
            SparqlRawOperation::Query(name, _) => name.clone(),
        }
    }

    pub fn text(&self) -> &str {
        match self {
            SparqlRawOperation::Query(_, text) => text.as_ref(),
        }
    }

    pub fn parse(&self) -> anyhow::Result<SparqlOperation<QueryName>> {
        match self {
            SparqlRawOperation::Query(query_name, query) => {
                let query = query.parse()?;
                Ok(SparqlOperation::Query(query_name.clone(), query))
            }
        }
    }
}

#[derive(Clone)]
pub enum SparqlOperation<QueryName> {
    Query(QueryName, Query),
}

impl<QueryName> SparqlOperation<QueryName> {
    pub fn query(&self) -> &Query {
        match self {
            SparqlOperation::Query(_, query) => query,
        }
    }

    pub async fn run(
        &self,
        store: &Store,
    ) -> anyhow::Result<(BenchmarkRun, QueryExplanation, usize)> {
        let start = datafusion::common::instant::Instant::now();

        let mut num_results = 0;
        let options = QueryOptions::default();
        let explanation = match &self {
            SparqlOperation::Query(_, q) => {
                let (result, explanation) =
                    store.explain_query_opt(q.clone(), options.clone()).await?;
                match result {
                    QueryResults::Boolean(_) => (),
                    QueryResults::Solutions(s) => {
                        let mut stream = s.into_record_batch_stream()?;
                        while let Some(s) = stream.next().await {
                            num_results += s?.num_rows();
                        }
                    }
                    QueryResults::Graph(mut g) => {
                        while let Some(t) = g.next().await {
                            t?;
                            num_results += 1;
                        }
                    }
                }
                explanation
            }
        };

        let duration = start.elapsed();
        Ok((BenchmarkRun { duration }, explanation, num_results))
    }
}

impl<QueryName: Clone> SparqlOperation<QueryName> {
    pub fn query_name(&self) -> QueryName {
        match self {
            SparqlOperation::Query(name, _) => name.clone(),
        }
    }
}
