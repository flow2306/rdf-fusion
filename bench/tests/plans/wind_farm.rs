use crate::plans::canonicalize_uuids;
use anyhow::Context;
use datafusion::physical_plan::displayable;
use insta::assert_snapshot;
use rdf_fusion::execution::sparql::{QueryExplanation, QueryOptions};
use rdf_fusion_bench::benchmarks::Benchmark;
use rdf_fusion_bench::benchmarks::windfarm::{
    NumTurbines, WindFarmBenchmark, WindFarmQueryName, get_wind_farm_raw_sparql_operation,
};
use rdf_fusion_bench::environment::{BenchmarkContext, RdfFusionBenchContext};
use rdf_fusion_bench::operation::SparqlRawOperation;
use std::path::PathBuf;

#[tokio::test]
pub async fn optimized_logical_plan_wind_farm() {
    for_all_explanations(|name, explanation| {
        assert_snapshot!(
            format!("{name} (Optimized)"),
            canonicalize_uuids(&explanation.optimized_logical_plan.to_string())
        )
    })
    .await;
}

#[tokio::test]
pub async fn execution_plan_wind_farm() {
    for_all_explanations(|name, explanation| {
        let string = displayable(explanation.execution_plan.as_ref())
            .indent(false)
            .to_string();
        assert_snapshot!(
            format!("{name} (Execution Plan)"),
            canonicalize_uuids(&string)
        )
    })
    .await;
}

async fn for_all_explanations(assertion: impl Fn(String, QueryExplanation) -> ()) {
    let benchmarking_context =
        RdfFusionBenchContext::new_for_criterion(PathBuf::from("./data"), 1);

    // Load the benchmark data and set max query count to one.
    let benchmark = WindFarmBenchmark::new(NumTurbines::N4);
    let benchmark_name = benchmark.name();
    let benchmark_context = benchmarking_context
        .create_benchmark_context(benchmark_name)
        .unwrap();

    let store = benchmark.prepare_store(&benchmark_context).await.unwrap();
    for query_name in WindFarmQueryName::list_queries() {
        let benchmark_name = format!("Wind Farm - {query_name}");
        let query = get_query_to_execute(&benchmark_context, query_name);

        let (_, explanation) = store
            .explain_query_opt(query.text(), QueryOptions::default())
            .await
            .unwrap();

        assertion(benchmark_name, explanation);
    }
}

fn get_query_to_execute(
    benchmark_context: &BenchmarkContext,
    query_name: WindFarmQueryName,
) -> SparqlRawOperation<WindFarmQueryName> {
    get_wind_farm_raw_sparql_operation(benchmark_context, query_name)
        .context("Could not list raw operations for Wind Farm benchmark. Have you prepared a wind-farm-4 dataset?")
        .unwrap()
}
