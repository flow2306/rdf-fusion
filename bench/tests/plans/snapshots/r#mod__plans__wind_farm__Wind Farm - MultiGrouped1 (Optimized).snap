---
source: bench/tests/plans/wind_farm.rs
expression: "&explanation.optimized_logical_plan.to_string()"
---
Projection: ENC_PT(site_label) AS site_label, ENC_PT(wtur_label) AS wtur_label, year AS year, month AS month, day AS day, hour AS hour, minute_10 AS minute_10, ENC_PT(<uuid>) AS val_prod_avg, ENC_PT(<uuid>) AS val_dir_avg, ENC_PT(<uuid>) AS val_speed_avg
  Aggregate: groupBy=[[site_label AS site_label, wtur_label AS wtur_label, ENC_PT(year) AS year, ENC_PT(month) AS month, ENC_PT(day) AS day, ENC_PT(hour) AS hour, ENC_PT(minute_10) AS minute_10]], aggr=[[AVG(ENC_TV(val_prod)) AS <uuid>, AVG(ENC_TV(val_dir)) AS <uuid>, AVG(ENC_TV(val_speed)) AS <uuid>]]
    Projection: site_label, wtur_label, val_speed, val_dir, val_prod, minute_10, hour, day, MONTH(__common_expr_4) AS month, YEAR(__common_expr_4) AS year
      Projection: ENC_TV(t) AS __common_expr_4, site_label, wtur_label, val_speed, val_dir, val_prod, minute_10, HOURS(__common_expr_5) AS hour, DAY(__common_expr_5) AS day
        Projection: __common_expr_9 AS __common_expr_5, site_label, wtur_label, val_speed, t, val_dir, val_prod, MUL(Union 9:10, FLOOR(DIV(MINUTES(__common_expr_9), Union 7:Some(<uuid>),36,16))) AS minute_10
          Projection: ENC_TV(t) AS __common_expr_9, site_label, wtur_label, val_speed, t, val_dir, val_prod
            Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.val_speed AS val_speed, lhs.t AS t, lhs.val_dir AS val_dir, lhs.val_prod AS val_prod
              Inner Join: lhs.dp_prod = rhs.dp_prod, lhs.t = rhs.t
                SubqueryAlias: lhs
                  Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.val_speed AS val_speed, lhs.t AS t, lhs.val_dir AS val_dir, lhs.dp_prod AS dp_prod, rhs.val_prod AS val_prod
                    Inner Join: lhs.dp_prod = rhs.dp_prod
                      SubqueryAlias: lhs
                        Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.val_speed AS val_speed, lhs.t AS t, lhs.val_dir AS val_dir, rhs.dp_prod AS dp_prod
                          Inner Join: lhs.ts_prod = rhs.ts_prod
                            SubqueryAlias: lhs
                              Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.val_speed AS val_speed, lhs.t AS t, lhs.val_dir AS val_dir, lhs.ts_prod AS ts_prod
                                Inner Join: lhs.ts_prod = rhs.ts_prod
                                  SubqueryAlias: lhs
                                    Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.val_speed AS val_speed, lhs.t AS t, lhs.val_dir AS val_dir, rhs.ts_prod AS ts_prod
                                      Inner Join: lhs.generator = rhs.generator
                                        SubqueryAlias: lhs
                                          Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.generator AS generator, lhs.val_speed AS val_speed, lhs.t AS t, lhs.val_dir AS val_dir
                                            Inner Join: lhs.ts_dir = rhs.ts_dir
                                              SubqueryAlias: lhs
                                                Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.generator AS generator, lhs.val_speed AS val_speed, lhs.t AS t, lhs.ts_dir AS ts_dir, lhs.val_dir AS val_dir
                                                  Inner Join: lhs.dp_dir = rhs.dp_dir, lhs.t = rhs.t
                                                    SubqueryAlias: lhs
                                                      Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.generator AS generator, lhs.val_speed AS val_speed, lhs.t AS t, lhs.ts_dir AS ts_dir, lhs.dp_dir AS dp_dir, rhs.val_dir AS val_dir
                                                        Inner Join: lhs.dp_dir = rhs.dp_dir
                                                          SubqueryAlias: lhs
                                                            Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.generator AS generator, lhs.val_speed AS val_speed, lhs.t AS t, lhs.ts_dir AS ts_dir, rhs.dp_dir AS dp_dir
                                                              Inner Join: lhs.ts_dir = rhs.ts_dir
                                                                SubqueryAlias: lhs
                                                                  Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.generator AS generator, lhs.val_speed AS val_speed, lhs.t AS t, rhs.ts_dir AS ts_dir
                                                                    Inner Join: lhs.weather = rhs.weather
                                                                      SubqueryAlias: lhs
                                                                        Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.generator AS generator, lhs.weather AS weather, lhs.val_speed AS val_speed, lhs.t AS t
                                                                          Inner Join: lhs.ts_speed = rhs.ts_speed
                                                                            SubqueryAlias: lhs
                                                                              Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.generator AS generator, lhs.weather AS weather, lhs.ts_speed AS ts_speed, lhs.val_speed AS val_speed, rhs.t AS t
                                                                                Inner Join: lhs.dp_speed = rhs.dp_speed
                                                                                  SubqueryAlias: lhs
                                                                                    Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.generator AS generator, lhs.weather AS weather, lhs.ts_speed AS ts_speed, lhs.dp_speed AS dp_speed, rhs.val_speed AS val_speed
                                                                                      Inner Join: lhs.dp_speed = rhs.dp_speed
                                                                                        SubqueryAlias: lhs
                                                                                          Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.generator AS generator, lhs.weather AS weather, lhs.ts_speed AS ts_speed, rhs.dp_speed AS dp_speed
                                                                                            Inner Join: lhs.ts_speed = rhs.ts_speed
                                                                                              SubqueryAlias: lhs
                                                                                                Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.generator AS generator, lhs.weather AS weather, rhs.ts_speed AS ts_speed
                                                                                                  Inner Join: lhs.weather = rhs.weather
                                                                                                    SubqueryAlias: lhs
                                                                                                      Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.generator AS generator, lhs.weather AS weather
                                                                                                        Inner Join: lhs.weather = rhs.weather
                                                                                                          SubqueryAlias: lhs
                                                                                                            Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.generator AS generator, rhs.weather AS weather
                                                                                                              Inner Join: lhs.weather_asp = rhs.weather_asp
                                                                                                                SubqueryAlias: lhs
                                                                                                                  Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.generator AS generator, rhs.weather_asp AS weather_asp
                                                                                                                    Inner Join: lhs.wtur = rhs.wtur
                                                                                                                      SubqueryAlias: lhs
                                                                                                                        Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.generator AS generator
                                                                                                                          Inner Join: lhs.generator = rhs.generator
                                                                                                                            SubqueryAlias: lhs
                                                                                                                              Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, rhs.generator AS generator
                                                                                                                                Inner Join: lhs.generator_asp = rhs.generator_asp
                                                                                                                                  SubqueryAlias: lhs
                                                                                                                                    Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, rhs.generator_asp AS generator_asp
                                                                                                                                      Inner Join: lhs.gensys = rhs.gensys
                                                                                                                                        SubqueryAlias: lhs
                                                                                                                                          Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys AS gensys
                                                                                                                                            Inner Join: lhs.gensys = rhs.gensys
                                                                                                                                              SubqueryAlias: lhs
                                                                                                                                                Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, rhs.gensys AS gensys
                                                                                                                                                  Inner Join: lhs.gensys_asp = rhs.gensys_asp
                                                                                                                                                    SubqueryAlias: lhs
                                                                                                                                                      Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, rhs.gensys_asp AS gensys_asp
                                                                                                                                                        Inner Join: lhs.wtur = rhs.wtur
                                                                                                                                                          SubqueryAlias: lhs
                                                                                                                                                            Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur
                                                                                                                                                              Inner Join: lhs.wtur = rhs.wtur
                                                                                                                                                                SubqueryAlias: lhs
                                                                                                                                                                  Projection: lhs.site_label AS site_label, lhs.wtur_label AS wtur_label, rhs.wtur AS wtur
                                                                                                                                                                    Inner Join: lhs.wtur_asp = rhs.wtur_asp
                                                                                                                                                                      SubqueryAlias: lhs
                                                                                                                                                                        Projection: lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, rhs.wtur_label AS wtur_label
                                                                                                                                                                          Inner Join: lhs.wtur_asp = rhs.wtur_asp
                                                                                                                                                                            SubqueryAlias: lhs
                                                                                                                                                                              Projection: lhs.site_label AS site_label, rhs.wtur_asp AS wtur_asp
                                                                                                                                                                                Inner Join: lhs.site = rhs.site
                                                                                                                                                                                  SubqueryAlias: lhs
                                                                                                                                                                                    Projection: lhs.site AS site, rhs.site_label AS site_label
                                                                                                                                                                                      Inner Join: lhs.site = rhs.site
                                                                                                                                                                                        SubqueryAlias: lhs
                                                                                                                                                                                          QuadPattern (?site <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://github.com/magbak/chrontext/rds_power#Site>)
                                                                                                                                                                                        SubqueryAlias: rhs
                                                                                                                                                                                          Filter: EBV(EQ(ENC_TV(site_label), Union 3:{value:Wind Mountain,language:}))
                                                                                                                                                                                            QuadPattern (?site <http://www.w3.org/2000/01/rdf-schema#label> ?site_label)
                                                                                                                                                                                  SubqueryAlias: rhs
                                                                                                                                                                                    QuadPattern (?site <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspect> ?wtur_asp)
                                                                                                                                                                            SubqueryAlias: rhs
                                                                                                                                                                              Filter: EBV(EQ(ENC_TV(wtur_label), Union 3:{value:A1,language:}))
                                                                                                                                                                                QuadPattern (?wtur_asp <http://www.w3.org/2000/01/rdf-schema#label> ?wtur_label)
                                                                                                                                                                      SubqueryAlias: rhs
                                                                                                                                                                        QuadPattern (?wtur <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspectNode> ?wtur_asp)
                                                                                                                                                                SubqueryAlias: rhs
                                                                                                                                                                  QuadPattern (?wtur <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://github.com/magbak/chrontext/rds_power#A>)
                                                                                                                                                          SubqueryAlias: rhs
                                                                                                                                                            QuadPattern (?wtur <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspect> ?gensys_asp)
                                                                                                                                                    SubqueryAlias: rhs
                                                                                                                                                      QuadPattern (?gensys <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspectNode> ?gensys_asp)
                                                                                                                                              SubqueryAlias: rhs
                                                                                                                                                QuadPattern (?gensys <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://github.com/magbak/chrontext/rds_power#RA>)
                                                                                                                                        SubqueryAlias: rhs
                                                                                                                                          QuadPattern (?gensys <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspect> ?generator_asp)
                                                                                                                                  SubqueryAlias: rhs
                                                                                                                                    QuadPattern (?generator <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspectNode> ?generator_asp)
                                                                                                                            SubqueryAlias: rhs
                                                                                                                              QuadPattern (?generator <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://github.com/magbak/chrontext/rds_power#GAA>)
                                                                                                                      SubqueryAlias: rhs
                                                                                                                        QuadPattern (?wtur <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspect> ?weather_asp)
                                                                                                                SubqueryAlias: rhs
                                                                                                                  QuadPattern (?weather <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspectNode> ?weather_asp)
                                                                                                          SubqueryAlias: rhs
                                                                                                            QuadPattern (?weather <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://github.com/magbak/chrontext/rds_power#LE>)
                                                                                                    SubqueryAlias: rhs
                                                                                                      QuadPattern (?weather <https://github.com/magbak/chrontext#hasTimeseries> ?ts_speed)
                                                                                              SubqueryAlias: rhs
                                                                                                QuadPattern (?ts_speed <https://github.com/magbak/chrontext#hasDataPoint> ?dp_speed)
                                                                                        SubqueryAlias: rhs
                                                                                          QuadPattern (?dp_speed <https://github.com/magbak/chrontext#hasValue> ?val_speed)
                                                                                  SubqueryAlias: rhs
                                                                                    Projection: dp_speed, t
                                                                                      Filter: EBV(GEQ(__common_expr_10, Union 10:{value:6379744601300.0000000000000000,offset:})) AND EBV(LEQ(__common_expr_10, Union 10:{value:6379749281300.0000000000000000,offset:}))
                                                                                        Projection: ENC_TV(t) AS __common_expr_10, dp_speed, t
                                                                                          QuadPattern (?dp_speed <https://github.com/magbak/chrontext#hasTimestamp> ?t)
                                                                            SubqueryAlias: rhs
                                                                              QuadPattern (?ts_speed <http://www.w3.org/2000/01/rdf-schema#label> "Windspeed")
                                                                      SubqueryAlias: rhs
                                                                        QuadPattern (?weather <https://github.com/magbak/chrontext#hasTimeseries> ?ts_dir)
                                                                SubqueryAlias: rhs
                                                                  QuadPattern (?ts_dir <https://github.com/magbak/chrontext#hasDataPoint> ?dp_dir)
                                                          SubqueryAlias: rhs
                                                            QuadPattern (?dp_dir <https://github.com/magbak/chrontext#hasValue> ?val_dir)
                                                    SubqueryAlias: rhs
                                                      Projection: dp_dir, t
                                                        Filter: EBV(GEQ(__common_expr_11, Union 10:{value:6379744601300.0000000000000000,offset:})) AND EBV(LEQ(__common_expr_11, Union 10:{value:6379749281300.0000000000000000,offset:}))
                                                          Projection: ENC_TV(t) AS __common_expr_11, dp_dir, t
                                                            QuadPattern (?dp_dir <https://github.com/magbak/chrontext#hasTimestamp> ?t)
                                              SubqueryAlias: rhs
                                                QuadPattern (?ts_dir <http://www.w3.org/2000/01/rdf-schema#label> "WindDirection")
                                        SubqueryAlias: rhs
                                          QuadPattern (?generator <https://github.com/magbak/chrontext#hasTimeseries> ?ts_prod)
                                  SubqueryAlias: rhs
                                    QuadPattern (?ts_prod <http://www.w3.org/2000/01/rdf-schema#label> "Production")
                            SubqueryAlias: rhs
                              QuadPattern (?ts_prod <https://github.com/magbak/chrontext#hasDataPoint> ?dp_prod)
                      SubqueryAlias: rhs
                        QuadPattern (?dp_prod <https://github.com/magbak/chrontext#hasValue> ?val_prod)
                SubqueryAlias: rhs
                  Projection: dp_prod, t
                    Filter: EBV(GEQ(__common_expr_12, Union 10:{value:6379744601300.0000000000000000,offset:})) AND EBV(LEQ(__common_expr_12, Union 10:{value:6379749281300.0000000000000000,offset:}))
                      Projection: ENC_TV(t) AS __common_expr_12, dp_prod, t
                        QuadPattern (?dp_prod <https://github.com/magbak/chrontext#hasTimestamp> ?t)
